File: rust/src/observability/prometheus.rs
//! Prometheus metrics exporter for links-queue.
//!
//! Provides Prometheus-compatible metrics export in text format,
//! following the Prometheus exposition format specification.

use super::metrics::{MetricsRegistry, QueueMetricsData, HistogramStats};
use std::collections::HashMap;
use std::sync::Arc;

/// Formats a metric name for Prometheus.
fn format_metric_name(name: &str, prefix: Option<&str>) -> String {
    let name = name
        .to_lowercase()
        .chars()
        .map(|c| if c.is_alphanumeric() || c == '_' || c == ':' { c } else { '_' })
        .collect::<String>();

    // Remove consecutive underscores and trim
    let name = name
        .split('_')
        .filter(|s| !s.is_empty())
        .collect::<Vec<_>>()
        .join("_");

    match prefix {
        Some(p) => format!("{}_{}", p, name),
        None => name,
    }
}

/// Escapes label values for Prometheus format.
fn escape_label_value(value: &str) -> String {
    value
        .replace('\\', "\\\\")
        .replace('"', "\\\"")
        .replace('\n', "\\n")
}

/// Formats labels as Prometheus label string.
fn format_labels(labels: &HashMap<String, String>) -> String {
    if labels.is_empty() {
        return String::new();
    }

    let pairs: Vec<String> = labels
        .iter()
        .map(|(k, v)| format!("{}=\"{}\"", k, escape_label_value(v)))
        .collect();

    format!("{{{}}}", pairs.join(","))
}

/// Prometheus exporter for links-queue metrics.
pub struct PrometheusExporter {
    registry: Arc<MetricsRegistry>,
    prefix: Option<String>,
    default_labels: HashMap<String, String>,
}

impl PrometheusExporter {
    /// Creates a new `PrometheusExporter`.
    #[must_use]
    pub fn new(registry: Arc<MetricsRegistry>) -> Self {
        Self {
            registry,
            prefix: None,
            default_labels: HashMap::new(),
        }
    }

    /// Sets the metric name prefix.
    #[must_use]
    pub fn with_prefix(mut self, prefix: impl Into<String>) -> Self {
        self.prefix = Some(prefix.into());
        self
    }

    /// Adds a default label.
    #[must_use]
    pub fn with_label(mut self, key: impl Into<String>, value: impl Into<String>) -> Self {
        self.default_labels.insert(key.into(), value.into());
        self
    }

    /// Formats a counter metric.
    fn format_counter(
        &self,
        name: &str,
        value: u64,
        help: &str,
        labels: &HashMap<String, String>,
    ) -> String {
        let metric_name = format_metric_name(name, self.prefix.as_deref());
        let mut all_labels = self.default_labels.clone();
        all_labels.extend(labels.clone());
        let label_str = format_labels(&all_labels);

        format!(
            "# HELP {} {}\n# TYPE {} counter\n{}{} {}",
            metric_name, help, metric_name, metric_name, label_str, value
        )
    }

    /// Formats a gauge metric.
    fn format_gauge(
        &self,
        name: &str,
        value: i64,
        help: &str,
        labels: &HashMap<String, String>,
    ) -> String {
        let metric_name = format_metric_name(name, self.prefix.as_deref());
        let mut all_labels = self.default_labels.clone();
        all_labels.extend(labels.clone());
        let label_str = format_labels(&all_labels);

        format!(
            "# HELP {} {}\n# TYPE {} gauge\n{}{} {}",
            metric_name, help, metric_name, metric_name, label_str, value
        )
    }

    /// Formats a histogram metric.
    fn format_histogram(
        &self,
        name: &str,
        stats: &HistogramStats,
        help: &str,
        labels: &HashMap<String, String>,
    ) -> String {
        let metric_name = format_metric_name(name, self.prefix.as_deref());
        let mut all_labels = self.default_labels.clone();
        all_labels.extend(labels.clone());

        let mut lines = vec![
            format!("# HELP {} {}", metric_name, help),
            format!("# TYPE {} histogram", metric_name),
        ];

        // Bucket values (cumulative)
        let mut cumulative = 0u64;
        for bucket in &stats.buckets {
            cumulative += bucket.count;
            let mut bucket_labels = all_labels.clone();
            bucket_labels.insert("le".to_string(), bucket.le.to_string());
            lines.push(format!(
                "{}_bucket{} {}",
                metric_name,
                format_labels(&bucket_labels),
                cumulative
            ));
        }

        // +Inf bucket
        let mut inf_labels = all_labels.clone();
        inf_labels.insert("le".to_string(), "+Inf".to_string());
        lines.push(format!(
            "{}_bucket{} {}",
            metric_name,
            format_labels(&inf_labels),
            stats.count
        ));

        // Sum and count
        let label_str = format_labels(&all_labels);
        lines.push(format!("{}_sum{} {}", metric_name, label_str, stats.sum));
        lines.push(format!("{}_count{} {}", metric_name, label_str, stats.count));

        lines.join("\n")
    }

    /// Exports queue metrics in Prometheus format.
    fn export_queue_metrics(&self, queue_name: &str, metrics: &QueueMetricsData) -> String {
        let mut labels = HashMap::new();
        labels.insert("queue".to_string(), queue_name.to_string());

        let mut parts = Vec::new();

        // Queue depth
        parts.push(self.format_gauge(
            "links_queue_depth",
            metrics.depth,
            "Current number of messages in the queue",
            &labels,
        ));

        // Throughput counters
        parts.push(self.format_counter(
            "links_queue_messages_enqueued_total",
            metrics.throughput.enqueued,
            "Total number of messages enqueued",
            &labels,
        ));
        parts.push(self.format_counter(
            "links_queue_messages_dequeued_total",
            metrics.throughput.dequeued,
            "Total number of messages dequeued",
            &labels,
        ));
        parts.push(self.format_counter(
            "links_queue_messages_acknowledged_total",
            metrics.throughput.acknowledged,
            "Total number of messages acknowledged",
            &labels,
        ));
        parts.push(self.format_counter(
            "links_queue_messages_rejected_total",
            metrics.throughput.rejected,
            "Total number of messages rejected",
            &labels,
        ));
        parts.push(self.format_counter(
            "links_queue_messages_dead_lettered_total",
            metrics.throughput.dead_lettered,
            "Total number of messages moved to dead letter queue",
            &labels,
        ));

        // Throughput rates
        parts.push(self.format_gauge(
            "links_queue_enqueue_rate",
            metrics.throughput.enqueue_rate as i64,
            "Current enqueue rate (messages/second)",
            &labels,
        ));
        parts.push(self.format_gauge(
            "links_queue_dequeue_rate",
            metrics.throughput.dequeue_rate as i64,
            "Current dequeue rate (messages/second)",
            &labels,
        ));

        // Latency histograms
        parts.push(self.format_histogram(
            "links_queue_enqueue_duration_milliseconds",
            &metrics.latency.enqueue,
            "Enqueue operation duration in milliseconds",
            &labels,
        ));
        parts.push(self.format_histogram(
            "links_queue_dequeue_duration_milliseconds",
            &metrics.latency.dequeue,
            "Dequeue operation duration in milliseconds",
            &labels,
        ));
        parts.push(self.format_histogram(
            "links_queue_processing_duration_milliseconds",
            &metrics.latency.processing,
            "Message processing duration in milliseconds",
            &labels,
        ));

        // Consumer lag and in-flight
        parts.push(self.format_gauge(
            "links_queue_consumer_lag",
            metrics.consumer_lag,
            "Number of messages waiting to be processed",
            &labels,
        ));
        parts.push(self.format_gauge(
            "links_queue_in_flight",
            metrics.in_flight,
            "Number of messages currently being processed",
            &labels,
        ));

        // Errors
        parts.push(self.format_counter(
            "links_queue_errors_total",
            metrics.errors,
            "Total number of errors",
            &labels,
        ));

        parts.join("\n\n")
    }

    /// Exports all metrics in Prometheus text format.
    #[must_use]
    pub fn export(&self) -> String {
        let mut parts = Vec::new();

        // Uptime
        parts.push(self.format_gauge(
            "links_queue_uptime_seconds",
            self.registry.uptime().as_secs() as i64,
            "Uptime of the links-queue process in seconds",
            &HashMap::new(),
        ));

        // Queue metrics
        let queues = self.registry.get_all_queue_metrics();
        for (name, metrics) in &queues {
            parts.push(self.export_queue_metrics(name, metrics));
        }

        parts.join("\n\n") + "\n"
    }

    /// Returns the content type for Prometheus metrics.
    #[must_use]
    pub fn content_type() -> &'static str {
        "text/plain; version=0.0.4; charset=utf-8"
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_format_metric_name() {
        assert_eq!(format_metric_name("my_metric", None), "my_metric");
        assert_eq!(format_metric_name("my-metric", None), "my_metric");
        assert_eq!(format_metric_name("my_metric", Some("app")), "app_my_metric");
    }

    #[test]
    fn test_format_labels() {
        let mut labels = HashMap::new();
        labels.insert("queue".to_string(), "test".to_string());
        labels.insert("env".to_string(), "prod".to_string());

        let result = format_labels(&labels);
        assert!(result.contains("queue=\"test\""));
        assert!(result.contains("env=\"prod\""));
    }

    #[test]
    fn test_escape_label_value() {
        assert_eq!(escape_label_value("hello"), "hello");
        assert_eq!(escape_label_value("hello\"world"), "hello\\\"world");
        assert_eq!(escape_label_value("hello\nworld"), "hello\\nworld");
    }

    #[test]
    fn test_exporter() {
        let registry = Arc::new(MetricsRegistry::new());
        let metrics = registry.get_queue_metrics("test");
        metrics.record_enqueue(Some(5.0));

        let exporter = PrometheusExporter::new(registry);
        let output = exporter.export();

        assert!(output.contains("# TYPE"));
        assert!(output.contains("links_queue_depth"));
        assert!(output.contains("queue=\"test\""));
    }

    #[test]
    fn test_content_type() {
        let ct = PrometheusExporter::content_type();
        assert!(ct.contains("text/plain"));
        assert!(ct.contains("version=0.0.4"));
    }
}

File: js/src/observability/prometheus.js
/**
 * Prometheus metrics exporter for links-queue.
 *
 * Provides Prometheus-compatible metrics export in text format,
 * following the Prometheus exposition format specification.
 *
 * @module observability/prometheus
 *
 * @see https://prometheus.io/docs/instrumenting/exposition_formats/
 * @see REQUIREMENTS.md - REQ-OBS-001 through REQ-OBS-004
 */

import { defaultRegistry } from './metrics.js';

/**
 * Formats a metric name for Prometheus.
 * Converts to lowercase and replaces invalid characters.
 *
 * @param {string} name - Original metric name
 * @returns {string} Prometheus-compatible name
 */
const formatMetricName = (name) =>
  name
    .toLowerCase()
    .replace(/[^a-z0-9_:]/g, '_')
    .replace(/_+/g, '_')
    .replace(/^_/, '')
    .replace(/_$/, '');

/**
 * Escapes label values for Prometheus format.
 *
 * @param {string} value - Label value
 * @returns {string} Escaped value
 */
const escapeLabelValue = (value) =>
  String(value)
    .replace(/\\/g, '\\\\')
    .replace(/"/g, '\\"')
    .replace(/\n/g, '\\n');

/**
 * Formats labels as Prometheus label string.
 *
 * @param {Object} labels - Label key-value pairs
 * @returns {string} Formatted label string
 */
const formatLabels = (labels) => {
  if (!labels || Object.keys(labels).length === 0) {
    return '';
  }
  const pairs = Object.entries(labels)
    .map(([key, value]) => `${key}="${escapeLabelValue(value)}"`)
    .join(',');
  return `{${pairs}}`;
};

/**
 * Prometheus exporter for links-queue metrics.
 */
export class PrometheusExporter {
  /**
   * Creates a new PrometheusExporter.
   *
   * @param {Object} [options] - Configuration options
   * @param {import('./metrics.js').MetricsRegistry} [options.registry] - Metrics registry to export
   * @param {string} [options.prefix] - Metric name prefix
   * @param {Object} [options.defaultLabels] - Default labels to add to all metrics
   */
  constructor(options = {}) {
    this._registry = options.registry ?? defaultRegistry;
    this._prefix = options.prefix ?? '';
    this._defaultLabels = options.defaultLabels ?? {};
  }

  /**
   * Gets the metrics registry.
   *
   * @returns {import('./metrics.js').MetricsRegistry} Registry
   */
  get registry() {
    return this._registry;
  }

  /**
   * Sets the metrics registry.
   *
   * @param {import('./metrics.js').MetricsRegistry} registry - New registry
   */
  set registry(registry) {
    this._registry = registry;
  }

  /**
   * Formats a counter metric for Prometheus.
   *
   * @param {string} name - Metric name
   * @param {number} value - Counter value
   * @param {string} help - Help text
   * @param {Object} [labels] - Additional labels
   * @returns {string} Prometheus format output
   * @private
   */
  _formatCounter(name, value, help, labels = {}) {
    const metricName = formatMetricName(
      this._prefix ? `${this._prefix}_${name}` : name
    );
    const allLabels = { ...this._defaultLabels, ...labels };
    const labelStr = formatLabels(allLabels);

    return [
      `# HELP ${metricName} ${help}`,
      `# TYPE ${metricName} counter`,
      `${metricName}${labelStr} ${value}`,
    ].join('\n');
  }

  /**
   * Formats a gauge metric for Prometheus.
   *
   * @param {string} name - Metric name
   * @param {number} value - Gauge value
   * @param {string} help - Help text
   * @param {Object} [labels] - Additional labels
   * @returns {string} Prometheus format output
   * @private
   */
  _formatGauge(name, value, help, labels = {}) {
    const metricName = formatMetricName(
      this._prefix ? `${this._prefix}_${name}` : name
    );
    const allLabels = { ...this._defaultLabels, ...labels };
    const labelStr = formatLabels(allLabels);

    return [
      `# HELP ${metricName} ${help}`,
      `# TYPE ${metricName} gauge`,
      `${metricName}${labelStr} ${value}`,
    ].join('\n');
  }

  /**
   * Formats a histogram metric for Prometheus.
   *
   * @param {string} name - Metric name
   * @param {Object} stats - Histogram stats from LatencyHistogram.getStats()
   * @param {string} help - Help text
   * @param {Object} [labels] - Additional labels
   * @returns {string} Prometheus format output
   * @private
   */
  _formatHistogram(name, stats, help, labels = {}) {
    const metricName = formatMetricName(
      this._prefix ? `${this._prefix}_${name}` : name
    );
    const allLabels = { ...this._defaultLabels, ...labels };
    const lines = [
      `# HELP ${metricName} ${help}`,
      `# TYPE ${metricName} histogram`,
    ];

    // Bucket values (cumulative)
    let cumulative = 0;
    for (const bucket of stats.buckets) {
      cumulative += bucket.count;
      const bucketLabels = { ...allLabels, le: String(bucket.le) };
      lines.push(
        `${metricName}_bucket${formatLabels(bucketLabels)} ${cumulative}`
      );
    }

    // +Inf bucket
    const infLabels = { ...allLabels, le: '+Inf' };
    lines.push(`${metricName}_bucket${formatLabels(infLabels)} ${stats.count}`);

    // Sum and count
    const labelStr = formatLabels(allLabels);
    lines.push(`${metricName}_sum${labelStr} ${stats.sum}`);
    lines.push(`${metricName}_count${labelStr} ${stats.count}`);

    return lines.join('\n');
  }

  /**
   * Formats a summary metric for Prometheus (for percentiles).
   *
   * @param {string} name - Metric name
   * @param {Object} stats - Histogram stats with percentiles
   * @param {string} help - Help text
   * @param {Object} [labels] - Additional labels
   * @returns {string} Prometheus format output
   * @private
   */
  _formatSummary(name, stats, help, labels = {}) {
    const metricName = formatMetricName(
      this._prefix ? `${this._prefix}_${name}` : name
    );
    const allLabels = { ...this._defaultLabels, ...labels };
    const lines = [
      `# HELP ${metricName} ${help}`,
      `# TYPE ${metricName} summary`,
    ];

    // Quantiles
    const quantiles = [
      { q: '0.5', value: stats.p50 },
      { q: '0.9', value: stats.p90 },
      { q: '0.95', value: stats.p95 },
      { q: '0.99', value: stats.p99 },
    ];

    for (const { q, value } of quantiles) {
      const quantileLabels = { ...allLabels, quantile: q };
      lines.push(`${metricName}${formatLabels(quantileLabels)} ${value}`);
    }

    // Sum and count
    const labelStr = formatLabels(allLabels);
    lines.push(`${metricName}_sum${labelStr} ${stats.sum}`);
    lines.push(`${metricName}_count${labelStr} ${stats.count}`);

    return lines.join('\n');
  }

  /**
   * Exports queue metrics in Prometheus format.
   *
   * @param {string} queueName - Queue name
   * @param {Object} metrics - Metrics from QueueMetrics.getMetrics()
   * @returns {string} Prometheus format output
   * @private
   */
  _exportQueueMetrics(queueName, metrics) {
    const labels = { queue: queueName };
    const parts = [];

    // Queue depth (gauge)
    parts.push(
      this._formatGauge(
        'links_queue_depth',
        metrics.depth,
        'Current number of messages in the queue',
        labels
      )
    );

    // Throughput counters
    parts.push(
      this._formatCounter(
        'links_queue_messages_enqueued_total',
        metrics.throughput.enqueued,
        'Total number of messages enqueued',
        labels
      )
    );
    parts.push(
      this._formatCounter(
        'links_queue_messages_dequeued_total',
        metrics.throughput.dequeued,
        'Total number of messages dequeued',
        labels
      )
    );
    parts.push(
      this._formatCounter(
        'links_queue_messages_acknowledged_total',
        metrics.throughput.acknowledged,
        'Total number of messages acknowledged',
        labels
      )
    );
    parts.push(
      this._formatCounter(
        'links_queue_messages_rejected_total',
        metrics.throughput.rejected,
        'Total number of messages rejected',
        labels
      )
    );
    parts.push(
      this._formatCounter(
        'links_queue_messages_dead_lettered_total',
        metrics.throughput.deadLettered,
        'Total number of messages moved to dead letter queue',
        labels
      )
    );

    // Throughput rates (gauges)
    parts.push(
      this._formatGauge(
        'links_queue_enqueue_rate',
        metrics.throughput.enqueueRate,
        'Current enqueue rate (messages/second)',
        labels
      )
    );
    parts.push(
      this._formatGauge(
        'links_queue_dequeue_rate',
        metrics.throughput.dequeueRate,
        'Current dequeue rate (messages/second)',
        labels
      )
    );

    // Latency histograms
    parts.push(
      this._formatHistogram(
        'links_queue_enqueue_duration_milliseconds',
        metrics.latency.enqueue,
        'Enqueue operation duration in milliseconds',
        labels
      )
    );
    parts.push(
      this._formatHistogram(
        'links_queue_dequeue_duration_milliseconds',
        metrics.latency.dequeue,
        'Dequeue operation duration in milliseconds',
        labels
      )
    );
    parts.push(
      this._formatHistogram(
        'links_queue_processing_duration_milliseconds',
        metrics.latency.processing,
        'Message processing duration in milliseconds',
        labels
      )
    );

    // Consumer lag and in-flight (gauges)
    parts.push(
      this._formatGauge(
        'links_queue_consumer_lag',
        metrics.consumerLag,
        'Number of messages waiting to be processed',
        labels
      )
    );
    parts.push(
      this._formatGauge(
        'links_queue_in_flight',
        metrics.inFlight,
        'Number of messages currently being processed',
        labels
      )
    );

    // Errors
    parts.push(
      this._formatCounter(
        'links_queue_errors_total',
        metrics.errors,
        'Total number of errors',
        labels
      )
    );

    return parts.join('\n\n');
  }

  /**
   * Exports all metrics in Prometheus text format.
   *
   * @returns {string} Prometheus exposition format output
   */
  export() {
    const allMetrics = this._registry.getAllMetrics();
    const parts = [];

    // Process info
    parts.push(
      this._formatGauge(
        'links_queue_uptime_seconds',
        Math.floor(allMetrics.uptime / 1000),
        'Uptime of the links-queue process in seconds'
      )
    );

    // Queue metrics
    for (const [queueName, metrics] of Object.entries(allMetrics.queues)) {
      parts.push(this._exportQueueMetrics(queueName, metrics));
    }

    // Global counters
    for (const [name, value] of Object.entries(allMetrics.counters)) {
      parts.push(this._formatCounter(name, value, `Global counter: ${name}`));
    }

    // Global gauges
    for (const [name, value] of Object.entries(allMetrics.gauges)) {
      parts.push(this._formatGauge(name, value, `Global gauge: ${name}`));
    }

    return `${parts.join('\n\n')}\n`;
  }

  /**
   * Returns the content type for Prometheus metrics.
   *
   * @returns {string} Content type header value
   */
  static contentType() {
    return 'text/plain; version=0.0.4; charset=utf-8';
  }

  /**
   * Static helper to format metrics from a registry.
   *
   * @param {import('./metrics.js').MetricsRegistry} registry - Metrics registry
   * @param {Object} [options] - Exporter options
   * @returns {string} Prometheus format output
   */
  static format(registry, options = {}) {
    const exporter = new PrometheusExporter({ ...options, registry });
    return exporter.export();
  }

  /**
   * Static helper to format metrics from QueueMetrics.
   *
   * @param {Object} metrics - Metrics from QueueMetrics.getMetrics()
   * @param {Object} [options] - Exporter options
   * @returns {string} Prometheus format output
   */
  static formatQueueMetrics(metrics, options = {}) {
    const exporter = new PrometheusExporter(options);
    return exporter._exportQueueMetrics(metrics.queue, metrics);
  }
}

/**
 * Express/Fastify compatible middleware for serving Prometheus metrics.
 *
 * @param {Object} [options] - Middleware options
 * @param {import('./metrics.js').MetricsRegistry} [options.registry] - Metrics registry
 * @param {string} [options.path] - Path to serve metrics on (default: /metrics)
 * @returns {Function} Middleware function
 */
export const metricsMiddleware = (options = {}) => {
  const registry = options.registry ?? defaultRegistry;
  const path = options.path ?? '/metrics';
  const exporter = new PrometheusExporter({ ...options, registry });

  return (req, res, next) => {
    // Support both Express-style (req.path) and raw HTTP (req.url)
    const requestPath =
      req.path ?? new URL(req.url, 'http://localhost').pathname;

    if (requestPath === path) {
      const metrics = exporter.export();
      res.setHeader('Content-Type', PrometheusExporter.contentType());
      res.end(metrics);
      return;
    }

    if (typeof next === 'function') {
      next();
    }
  };
};
